reconnect_conn.go

```diff
@@ -0,0 +1,148 @@
+package sloglogstash
+
+import (
+	"errors"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type ReconnectConn struct {
+	dialer func() (net.Conn, error) // Function to dial new connection
+	conn   net.Conn                 // Current active connection
+	mu     sync.Mutex               // Mutex to protect conn
+
+	retryDelay time.Duration // Delay between reconnect attempts
+	maxRetries int           // Maximum number of reconnect attempts
+
+	closing   atomic.Bool // Indicates if the connection is closing/closed (atomic)
+	closeOnce sync.Once   // Ensures close is only performed once
+}
+
+func NewReconnectConn(dialer func() (net.Conn, error), retryDelay time.Duration, maxRetries int) (*ReconnectConn, error) {
```

Comment:
I wonder if we should dial lazily, on first error message ?

Because it could be blocking your main thread if logstash is unavailable.

I would move maxRetries in a setter, since it will be infinite most of the time.

WDYT?

Reply:
Good point. I changed it so the connection is first dialed during first write. Also added setter for maxRetries with 0 (unlimited) being the default.